README.md
# 🐉 DragonPaint 🐉 

DragonPaint.lua -- an app to help plan coloring of dragons for 'Dragon Adventures' game.  

This program runs with the Löve2D framework.  
(Developed with Love2D version 11.4) 

Concept and art by EveeChu  

(Most info will probably be in the Wiki)  


ChangeLog.md
# ChangeLog.md  
*(times read bottom to top) (use trailing doublespaces)*  



auto-format doc 
doing texture work on dev2  
merge dev1 to main (Release 1)  
-- 7/13 -- 


eksha start adding textures  
-- 7/12 -- 

4:20pm - eksha - use Lua dot notation, rather than array dereferences. 
3:00pm - change arrays to descriptive names & calculate image center.  
1:20pm - create d1_kilo branch  

1:15pm - create dev1 branch  
1:00pm - creating ChangeLog.md   
-- 7/10/22 --  


(Previous history only in git and code comments.)  
-- 7/9/22 --  


strict.lua
-- from: https://github.com/deepmind/strict/blob/master/strict.lua

local IGNORED_WRITES = {}
local IGNORED_READS = {
    qt=true,
    _PROMPT=true,
    _PROMPT2=true,
    writeObjects=true,
    arg=true,
}

-- Raises an error when an undeclared variable is read.
local function guardGlobals()
    assert(getmetatable(_G) == nil, "another global metatable exists")

    -- The detecting of undeclared vars is discussed on:
    -- http://www.lua.org/pil/14.2.html
    -- http://lua-users.org/wiki/DetectingUndefinedVariables
    setmetatable(_G, {
        __newindex = function (table, key, value)
            if not IGNORED_WRITES[key] then
                local info = debug.getinfo(2, "Sl")
                io.stderr:write(string.format(
                    "strict: %s:%s: write to undeclared variable: %s\n",
                    tostring(info.short_src), tostring(info.currentline), key))
            end
            rawset(table, key, value)
        end,
        __index = function (table, key)
            if IGNORED_READS[key] then
                return
            end
            error("attempt to read undeclared variable "..key, 2)
        end,
    })

    local origRequire = require
    function require(modname)
        IGNORED_WRITES[modname] = true
        return origRequire(modname)
    end
end

guardGlobals()


zipLove.ps1
# Script (powershell) to package DragonPaint files as a .love file 

# 9/5/22 - update to include ColorListSelector library. 
# 9/7/22 - fix lack of subfolders.  just COPY all the files I need, then zip there, then delete. 

write-host "`n"

# create (temporary) release folders:
New-Item -Path '.\tmp_Delete\DragonPaint' -ItemType Directory 
New-Item -Path '.\tmp_Delete\DragonPaint\lib' -ItemType Directory 
New-Item -Path '.\tmp_Delete\DragonPaint\lib\ColorListSelector' -ItemType Directory 

write-host "`n"

# --------------------------------------------------------------------- 
# Copy all the files needed for build, into their proper subfolders 

$copies = @{
    LiteralPath  = ".\main.lua", ".\DragonPaint.lua", ".\conf.lua", ".\ColorListConfig.lua"
    Destination  = ".\tmp_Delete\DragonPaint\"
}
#Copy-Item -WhatIf @copies   # show Test output ('WhatIf')
Copy-Item @copies 


$copies = @{
    LiteralPath  = ".\images\"
    Destination  = ".\tmp_Delete\DragonPaint\"
}
#Copy-Item -WhatIf @copies -Recurse
Copy-Item @copies -Recurse  # have to use -Recurse, or it won't inclue the files... 


$copies = @{
    LiteralPath  = ".\lib\strict.lua" 
    Destination  = ".\tmp_Delete\DragonPaint\lib\"
}
#Copy-Item -WhatIf @copies 
Copy-Item @copies 


$copies = @{
    LiteralPath  = ".\lib\ColorListSelector\ColorListSelector.lua", ".\lib\ColorListSelector\ColorListConfig.lua" 
    Destination  = ".\tmp_Delete\DragonPaint\lib\ColorListSelector\"
}
#Copy-Item -WhatIf @copies 
Copy-Item @copies 

write-host "`n"


# --------------------------------------------------------------------- 
# .zip all necessary project files together: 
# old: Compress-Archive -LiteralPath .\main.lua, .\DragonPaint.lua, .\conf.lua, .\strict.lua, .\images -CompressionLevel Optimal -DestinationPath .\DragonPaint.zip 
#
# multi-line style (more readable)
$compress = @{
    #LiteralPath = ".\tmp_Delete\DragonPaint\" # (no.. don't use 2 subfolders deep.) 
    Path = ".\tmp_Delete\DragonPaint\*"
    CompressionLevel = "Optimal"
    DestinationPath  = ".\DragonPaint.zip"
}
Compress-Archive -WhatIf @compress  # show Test output ('WhatIf')
Compress-Archive @compress 


# vv OLD vv 
# The single 'path' string below no longer works after library subdirectories were added; 
# it moves all the subdir files to the Root level... duh! 
$compress = @{
    LiteralPath      = ".\main.lua", ".\DragonPaint.lua", ".\conf.lua", ".\images", ".\lib\strict.lua", 
    ".\lib\ColorListSelector\ColorListSelector.lua", ".\lib\ColorListSelector\ColorListConfig.lua", ".\ColorListConfig.lua"
    CompressionLevel = "Optimal"
    DestinationPath  = ".\DragonPaint.zip"
}
#Compress-Archive -WhatIf @compress  # show Test output ('WhatIf')
#Compress-Archive @compress 
# ^^ OLD ^^ 

write-host "`n" 


# --------------------------------------------------------------------- 
# Rename .zip to .love 
Rename-Item -WhatIf -Path .\DragonPaint.zip -NewName .\DragonPaint.love   # show Test output  
Rename-Item -Path .\DragonPaint.zip -NewName .\DragonPaint.love


write-host "`n" 

# delete tmp folder:  '.\tmp_Delete\' 
Remove-Item '.\tmp_Delete' -Recurse   # (use -Recurse if you Don't want to be prompted for confirmation)
#Remove-Item '.\tmp_Delete'


ColorListConfig.lua
-- ColorListConfig -- LOCAL VERSION for DragonPaint.lua --

-- contains the configurable data to be used by --> ColorListSelector.lua

-- kmk todo:  create a setting for X position of the color menu...

-- set the size of the Color buttons in the complete color list
local buttonHeight = 30 -- default 30
local buttonSpacing = 6 -- default 6

local UIcanvasData = {
    --[[
        The "full screen" of the UI (small smartphone size by default) 
        Everything specified in ColorListConfig.lua is drawn on THIS canvas. 
        Normally, you'll just want to make this canvas the same dimensions 
        that your overall "app" window is.  It will be drawn on top of your app 
        (but its background is transparent, so your other content shows through.)
    ]]
    width = 640 * 1,
    height = 360 * 1,
    -- Normally the canvas background color should be left as transparent black,
    -- but sometimes it's useful set a slight color to be able to see the boundaries.
    -- bgColor = {0, 0, 1, 0.2},
    bgColor = { 0, 0, 0, 0 },
}

local colorsCanvasData = { -- config for the colorsCanvas
    --[[
        Here you can set the position & width of the Color Selection window. 
    --]]
    xPos = 1050, -- x position on the app screen
    yPos = 0, -- initial y position (this value will change when user scrolls the window)
    width = 200, -- ('height' is calculated based on list length)
    speed = 8, -- scroll speed for things like arrow keys
    -- Normally the canvas background color should be left as transparent black,
    -- but sometimes it's useful set a slight color to be able to see the boundaries.
    -- bgColor = { 0, 1, 0, 0.3 },
    -- bgColor = { 0, 0, 0, 0 },
    bgColor = { .1, .1, .1, .5 },

    -- kmk todo: move these into code-only since they're not really "configurable"
    height = 500, -- (this was just an initial value for development, it's actually recalculated in createColorsCanvas() )
    yStartDr = 0, -- the y position of the canvas at the start of a Drag
    active = false,
}

--[[ Below, you can define as many Buttons as you like, and add them to the button list. 
     Each button will trigger the Color Selector to appear. 
     You can 
--]]

-- Color-Select Buttons -- trigger the color select menu to appear, and show the currently chosen color
local csButton1 = { -- rectangle button with a text label
    text = "Primary", -- text label on the button
    -- kmk todo: specify font size, or compute it, or give user a hook to set it?
    x = 50,
    y = 140,
    width = 300,
    height = 80,
    color = { .6, .4, .4 }, -- default starting color

    -- kmk todo: remove these from here... add it in code/init.
    color_listNumber = 1, -- kmk, actually, we could just save list numbers for everything, rather than colors...
    color_previous = { .6, .4, .4 },
}

local csButton2 = {
    text = "Secondary",
    x = 50,
    y = 250,
    width = 300,
    height = 80,
    color = { .4, .4, .6 }, -- default starting color

    color_listNumber = 1,
    color_previous = { .4, .4, .6 },
}

local buttonList = { csButton1, csButton2 } -- list of Color-Select Trigger Buttons (rectangle ~objects) to place on screen


---------------------------------------------------------------------------------------------
-- kmk: I probably should have put the {RGB}s in a table, so they'd be easier to assign...
-- maybe each colorEntry = { name="", ColorRGB = {r,g,b} } so access is more "self documenting"
local colorList = { -- DEFAULT "test" color list... it's ok for later code to overwrite this.
    { "Test Colors", 1, 1, 1 },
    { "Red", 1, 0, 0 },
    { "Yellow", 1, 1, 0 },
    { "Magenta", 1, 0, 1 },
    { "Green", 0, 1, 0 },
    { "Cyan", 0, 1, 1 },
    { "Blue", 0, 0, 1 },
    { "dRed", .6, 0, 0 },
    { "dYellow", .6, .6, 0 },
    { "dMagenta", .6, 0, .6 },
    { "dGreen", 0, .6, 0 },
    { "dCyan", 0, .6, .6 },
    { "dBlue", 0, 0, .6 },
}

--[[
    kmk: currently the canvas size of the list below is too big for android to handle! 
    I guess making a giant canvas is not very memory efficient, but re-implement it to 
    generate the buttons dynamically as they scroll would take some effort... so do that later. 
    (implement it like side-scolling platformers do?)
    Determine what the max is for now. 
--]]

-- 3 hex values. need to convert to 3 RGB
local colorHexList = {
    'FFFFFF', 'White',
    'F7F9F9', 'Snowflake',
    'EAEDEF', 'Whisp',
    'D0CFD7', 'Whale',
    'AFAFAF', 'Silver',
    '888F8D', 'Gravel',
    '9C8E8D', 'Felt',
    '6A7185', 'Bluesteel',
    '636268', 'Stone',
    '5A6050', 'Tin',
    '545365', 'Spirit',
    '595451', 'Gloom',
    '4C4C4C', 'Coal',
    '4D484F', 'Gabbro',
    '413C40', 'Asphalt',
    '3B3736', 'Ash',
    '332D25', 'Basalt',
    '302722', 'Scoria',
    '1A1A1B', 'Black',
    '0E1011', 'Pitch',
    '1F1A23', 'Night',
    '22263D', 'Depth',
    '471A43', 'Blackberry',
    '4C2A4F', 'Berry',
    '553348', 'Loulou',
    '6E235D', 'Lilac',
    '863290', 'Grape',
    '9778BE', 'Petal',
    '7F6195', 'Satin',
    '5C415D', 'Haunted',
    '735B77', 'Ghost',
    '8E7F9E', 'Lavender',
    'A794B2', 'Amethyst',
    'AA96A6', 'Dart',
    'E1CDFE', 'Pansy',
    'CCA4E0', 'Bubble',
    'DA4FFF', 'Plum',
    '9C50D3', 'Purple',
    '993BD1', 'Eggplant',
    '7930B5', 'Midnight',
    '5317B5', 'Urchin',
    '4D2C89', 'Jelly',
    '3F2B66', 'Smog',
    '0D0A5B', 'Sapphire',
    '2B0D88', 'Angler',
    '2D237A', 'Bluebell',
    '484AA1', 'Aster',
    '525195', 'Smoke',
    '4866D5', 'Uranus',
    '757ADB', 'Rain',
    '7895C1', 'Stream',
    '444F69', 'Harpy',
    '324BA9', 'Blue',
    '212B5F', 'Denim',
    '013485', 'Morpho',
    '023AE2', 'Raindrop',
    '1C51E7', 'Marine',
    '2F83FF', 'Ocean',
    '6394DD', 'Drip',
    '76A8FF', 'Cool',
    'AEC8FF', 'Sky',
    '89A4C0', 'Cloud',
    '556979', 'Aluminum',
    '2F4557', 'Iron',
    '263746', 'Dream',
    '0D1E25', 'Abyss',
    '0B2D46', 'Trench',
    '0A3D67', 'Twilight',
    '094869', 'Mountain',
    '2B768F', 'Azure',
    '0086CE', 'Shell',
    '00B4D5', 'Cerulean',
    'B3E1F1', 'Winter',
    '91FFF7', 'Glow',
    '00FFF1', 'Cyan',
    '3CA2A4', 'Turquoise',
    '3A8684', 'History',
    '8DBCB4', 'Spruce',
    '72C4C4', 'Water',
    '9AEAEF', 'Glass',
    'E2FFE6', 'Pistachio',
    'B3FFD8', 'Dolphin',
    '9AFFC7', 'Mint',
    'B2E2BD', 'Seafoam',
    'A6DBA7', 'Caterpillar',
    '61AB89', 'Jade',
    '148E67', 'Spearmint',
    '1F565D', 'Essence',
    '233253', 'Rainforest',
    '153F4B', 'Seaweed',
    '114D41', 'Algae',
    '1F483A', 'Forest',
    '005D48', 'Hydra',
    '20603F', 'Emerald',
    '236825', 'Shamrock',
    '66903C', 'Pear',
    '1E361A', 'Jungle',
    '1E2716', 'Swamp',
    '1F281D', 'Root',
    '425035', 'Snake',
    '51684C', 'Camo',
    '516760', 'Scale',
    '687F67', 'Ivy',
    '97AF8B', 'Mantis',
    'A7B08C', 'Micah',
    '9BFF9D', 'Pea',
    '03ff7d', 'Synthesizer',
    '87E34D', 'Malachite',
    '7ECE73', 'Fern',
    '7BBD5D', 'Stem',
    '629C3F', 'Green',
    '567C34', 'Grass',
    '8ECE56', 'Cactus',
    'A5E32D', 'Leaf',
    'C6FF00', 'Toxin',
    'CDFE6C', 'Uranium',
    '9FFF00', 'Corrosion',
    'E8FCB4', 'Peridot',
    'D1E572', 'Cabbage',
    'B4CD3D', 'Chartreuse',
    'A9A032', 'Prehistoric',
    '828335', 'Alligator',
    '697135', 'Olive',
    '4B4420', 'Murk',
    '7E7645', 'Bark',
    'C18E1B', 'Amber',
    'BEA55D', 'Sponge',
    'D1B045', 'Haze',
    'D1B300', 'Swallowtail',
    'FFE63B', 'Lemon',
    'F9E255', 'Wasp',
    'F7FF6F', 'Yolk',
    'FFEC80', 'Banana',
    'FDD68B', 'Honey',
    'FDE9AC', 'Squash',
    'EDE8B0', 'Sanddollar',
    'FFFDEA', 'Mellow',
    'FDF1E1', 'Lychee',
    'FFEFDC', 'Creme',
    'F7DEBF', 'Pelt',
    'FFD297', 'Ivory',
    'F6BF6C', 'Peanut',
    'F2AD0C', 'Gold',
    'FFB53C', 'Marigold',
    'FA912B', 'Apricot',
    'FF8500', 'Poppy',
    'FF984F', 'Yam',
    'FFA147', 'Orange',
    'FFB576', 'Peach',
    'FCC4AD', 'Silt',
    'F0B392', 'Sahara',
    'D5602B', 'Saffron',
    'B2560D', 'Bronze',
    'B24407', 'Sandstone',
    'FF5500', 'Carrot',
    'EF5C23', 'Fire',
    'FF6841', 'Pumpkin',
    'FF7360', 'Sunrise',
    'C15A39', 'Cinnamon',
    'C47149', 'Caramel',
    'B27749', 'Acorn',
    '9A7B4F', 'Tortilla',
    'C3996F', 'Hide',
    'CABBA2', 'Beige',
    '827A64', 'Pine',
    '6D675B', 'Soil',
    '564D48', 'Coffee',
    '3C3030', 'Cocoa',
    '766259', 'Chocolate',
    '977B6C', 'Cappuccino',
    'BFA18F', 'Beach',
    '8A6059', 'Gingerbread',
    '7A4D4D', 'Maple',
    '774840', 'Hazel',
    '6B3C34', 'Coconut',
    '603E3D', 'Clay',
    '57372C', 'Sable',
    '432711', 'Penny',
    '301E1A', 'Umber',
    '22110A', 'Brownie',
    '2F1B1B', 'Birch',
    '5A4534', 'Feldspar',
    '72573A', 'Walnut',
    '855B33', 'Grain',
    '91532A', 'Ginger',
    '90553A', 'Starfish',
    '8E5B3F', 'Brown',
    '563012', 'Slate',
    '7B3C1D', 'Auburn',
    'A44B28', 'Copper',
    '8B3220', 'Rust',
    'BA311C', 'Tomato',
    'E22D18', 'Vermillion',
    'CE000D', 'Pepper',
    'AA0024', 'Cherry',
    '850012', 'Crimson',
    '7A0E1E', 'Ruby',
    '581014', 'Garnet',
    '2D0102', 'Sanguine',
    '451717', 'Blood',
    '652127', 'Rose',
    '8C272D', 'Cranberry',
    'C1272D', 'Redwood',
    'DF3236', 'Strawberry',
    'fc6d68', 'Fruit',
    'B13A3A', 'Carmine',
    'A12928', 'Cerise',
    '9A534D', 'Brick',
    'CC6F6F', 'Coral',
    'FEA0A0', 'Blush',
    'FFE2E6', 'Macaron',
    'FFB7B4', 'Sakura',
    'FEA1B3', 'Flamingo',
    'FFE5E5', 'Peony',
    'FF839B', 'Ribbon',
    'c67a80', 'Charm',
    'EB799A', 'Candy',
    'FB5E79', 'Bubblegum',
    'DB518D', 'Watermelon',
    'E934AA', 'Magenta',
    'E7008B', 'Fuschia',
    'cb0381', 'Tulip',
    'aa004c', 'Rubellite',
    '8A024A', 'Raspberry',
    '4D0F28', 'Syrah',
    '9C4975', 'Mauve',
    'E77FBF', 'Gum',
    'E5A9FF', 'Quartz',
    'E8CCFF', 'Confetti',
    'FFD6F6', 'Petalite',
    'FBEDFA', 'Pearl',
}


-- EXPORT Config data (for use by ColorListSelector.lua) --
--[[
    Store the Config Data needed externally (by ColorListSelector.lua) in a GLOBAL called 'CLSconfig' 
    (yeah, it's ugly to use a global, but I don't think Lua 5.1 could pass data as 
    extra arguments in require() calls yet.)
--]]
CLSconfig = { -- This is a GLOBAL var for now --
    UIcanvasData = UIcanvasData,
    colorsCanvasData = colorsCanvasData,
    buttonHeight = buttonHeight,
    buttonSpacing = buttonSpacing,
    colorList = colorList,
    colorHexList = colorHexList,
    buttonList = buttonList,
}


--[[ Return the module Data objects to be referenced by ColorListSelector:

    (This is in case future versions of Love2D use a later version of Lua that
    supports passing arguments in require() and this won't have to be a global anymore.)
--]]
return CLSconfig


DragonPaint.lua
local thisFile = "DragonPaint.lua"
local gameTitle = "DragonPaint"
local dpVersion = "0.08"
local dpVdate = "9/3/22"
print("[" .. thisFile .. "] version " .. dpVersion .. "\n")

local strict = require "lib.strict"
--print(strict.version)


----------------------------------
-- Prepare ColorListSelector (CLS)
-- First, get the CLS Config data (from wherever you keep the file)
-- User app can choose to use the default (lib) config, or call a custom one of their own.
require "lib.ColorListSelector.ColorListConfig" -- (the default library config)
-- require "ColorListConfig" -- (a local 'custom' config)

-- (ColorListConfig creates the global 'CLSconfig', which is mostly used by
--  ColorListSelector, but user code could potentially modify it if needed.)
local CLS = require "lib.ColorListSelector.ColorListSelector" -- Then get a handle to the API functions for CLS.
----------------------------------


local lg = love.graphics -- handy abbreviation

local mobile = false -- detect if running on an android phone

-- get timestamp of last save of thisFile (to use as a "minor version number")
local fileInfo = love.filesystem.getInfo(thisFile)
local verMinutes = math.floor((fileInfo.modtime - 1659000000) / 60)
print(thisFile .. " minutes since first version " .. verMinutes)
--(this is approximately the minutes since the first mobile version)


--[[ Design Notes ------------------------------------
This program requires the Love2D framework. 
(See the GitHub Wiki for more info.)

! Remember to run Auto-Format before major checkins (or right after), so random diffs don't appear. 

Let the DraImgList be Global. 
--]] -------------------------------------------------

--[[ --- To Do --- 
- REMEMBER to do main development for MOBILE dimensions 

-- Color list overlaps buttons, if you click on a color that overlaps a button, button and color are selected at once.
-- Streching screen cuts off selector buttons

- write function to saveColor(1), and revertColor(1) to abstract away the different implementations 
- use remember colors by their List index, not by rgb values. 

- support BIG lists: 
-- implement pageUp/Dn buttons 
-- dynamically generate buttons 1 page at a time. 
-- to a ~thumbindex scroll? 
-- do a 2nd zoomed out list? 


- change 'none' to '[smooth] plastic'
- add an 'exit' button? 
- have a lightly textured backgroud, always moving (clouds?)

- make a "credits" page.. spash screen or exit screen saying who the devs/owners 
are and what the "official" source site is.  (because games can be copied around 
and leave no trace of their origin.. other than their name.)
--]] -------------------------------------------------


-- dragon image info
local DraImgList = {} -- a table of 4 body regions.
-- kmk update: current COLORS will just be stored in the CLS ~Button objects.

local iPrimary = 1 -- some constants to make the code more readable
local iSecondary = 2
local iTertiary = 3

local function getRandomColor() -- returns a color as a Table

    -- math.random(lower, upper) generates integer numbers between lower and upper (both inclusive).
    local rb = math.random(0, 255)
    local gb = math.random(0, 255)
    local bb = math.random(0, 255)
    local ab = 255
    print("rgb color = ", rb, gb, bb)

    local colorT = { love.math.colorFromBytes(rb, gb, bb, ab) } -- pack into a Table...
    -- print("color = ", unpack(color))
    -- NOTE: Love2D/LuaJIT/Lua5.1 does not appear to have table.unpack()... it just has "unpack()"

    return colorT
end


-- kmk this can be DELETED once all 3 buttons are defined
local function randomColorDraImgList()
    print ""
    DraImgList.primary.color = getRandomColor()
    DraImgList.secondary.color = getRandomColor()
    DraImgList.tertiary.color = getRandomColor()
end


local function materialDragonImageList()

    -- having an indexed table makes random picking easier than elseifs
    local materials = { 'none', 'glass', 'sand' }

    local matfile = {}
    matfile.none = {}
    matfile.glass = {}
    matfile.sand = {}

    -- list of filenames makes it easy to add new material files
    matfile['none'].primary = "images/transparentPixel.png"
    matfile['none'].secondary = "images/transparentPixel.png"
    matfile['none'].tertiary = "images/transparentPixel.png"

    matfile['glass'].primary = "images/simpleAgricosPrimaryGlass.png"
    matfile['glass'].secondary = "images/simpleAgricosSecondaryGlass.png"
    matfile['glass'].tertiary = "images/simpleAgricosTertiaryGlass.png"

    matfile['sand'].primary = "images/simpleAgricosPrimarySand.png"
    matfile['sand'].secondary = "images/simpleAgricosSecondarySand.png"
    matfile['sand'].tertiary = "images/simpleAgricosTertiarySand.png"


    -- pick materials with random()
    local matPick = materials[math.random(#materials)]
    --print(matPick)
    --print(matfile[matPick].primary)
    DraImgList.primary.matImage = lg.newImage(matfile[matPick].primary)

    matPick = materials[math.random(#materials)]
    --print(matfile[matPick].secondary)
    DraImgList.secondary.matImage = lg.newImage(matfile[matPick].secondary)

    matPick = materials[math.random(#materials)]
    --print(matfile[matPick].tertiary)
    DraImgList.tertiary.matImage = lg.newImage(matfile[matPick].tertiary)
end


local function loadDragonImageList() -- store image regions, materials, & colors

    local drImage = nil -- dragon Image var
    local rColorT = {} -- table to hold 3 parts of an RGB color

    -- [] maybe make a "config file" of available images?
    local imfile_old = {
        outlines  = "images/dragontestLines.png",
        primary   = "images/dragontestPrimary.png",
        secondary = "images/dragontestSecondary.png",
        tertiary  = "images/dragontestTertiary.png"
    }

    local imfile = {
        outlines  = "images/simpleAgricosLines.png",
        primary   = "images/simpleAgricosPrimary.png",
        secondary = "images/simpleAgricosSecondary.png",
        tertiary  = "images/simpleAgricosTertiary.png",
    }

    -- color won't matter for the Outline, because it's black, but
    -- just in case anything in the outline image is filled white (e.g. eyes)
    -- the color for the "outline" image should be set to white.
    rColorT = { 1, 1, 1 }
    DraImgList.outlines = { image = lg.newImage(imfile.outlines), color = rColorT } -- kmk: move this color set to where the others are
    DraImgList.primary = { image = lg.newImage(imfile.primary) } -- (have to use ~constructor syntax for first assignment)
    DraImgList.secondary = { image = lg.newImage(imfile.secondary) }
    DraImgList.tertiary = { image = lg.newImage(imfile.tertiary) }

    randomColorDraImgList()
    materialDragonImageList()
end


function love.load() -- This is where Love2D does it's FIRST initialization.
    --create game window if not already in conf.lua
    --love.window.setMode(640, 360, {resizable=true, minwidth=400, minheight=300} )

    if love.system.getOS() == "Android" then
        mobile = true
    else
        mobile = false
        local desktopScale = 1 -- 1 = mobile, 2 = computer
        love.window.setMode(640 * desktopScale, 360 * desktopScale, { resizable = true })
        -- (fix? to get rid of the blink & resize, could set window to nil in conf.lua and only create them here)
    end

    love.window.setTitle(gameTitle .. "  v" .. dpVersion .. "-" .. verMinutes)
    --lg.setBackgroundColor(.7, .8, .9)  -- pale blue
    lg.setBackgroundColor(.9, .9, .8) -- pale tan
    lg.setColor(0, 0, 0)

    math.randomseed(os.time()) -- (lua5.1 always returns nil)

    loadDragonImageList()

    -- TEMPORARY --
    -- print ""
    -- print("window width/2 = " .. math.floor(lg.getWidth() / 2))
    -- print("window height/2 = " .. math.floor(lg.getHeight() / 2))

    -- local img = DraImgList.outlines.image
    -- print("0.5 image width/2 = " .. math.floor(img:getWidth() / 2 * 0.5))
    -- print("0.5 image height/2 = " .. math.floor(img:getHeight() / 2 * 0.5))
    -- print ""

    --[[ -- CLS Color List setup -- 
        At this point, you can either proceed with CLSconfig.colorList 
        as-is if you already have one defined in the correct format. 

        Otherwise, if you have a Color List in the simpler Flat CSV 
        format, you can call CLS.formatColorList() to convert it to 
        the runtime format. 
    --]]
    -- (Overwrite the default colorList with a reformatted Flat list)
    CLSconfig.colorList = CLS.formatColorList(CLSconfig.colorListData)

    -- Otherwise, if you have a Color List in any Non-Standard format, you
    -- can call your own custom 'formatColorList()' function to normalize it.
    CLSconfig.colorList = CLS.formatColorHexList(CLSconfig.colorHexList)

    CLS.load() -- let ColorListSelector do its initialization
end


function love.update(dt) -- Love2D calls this 60 times per second.
    CLS.update(dt)
end


-- change a "percentage" coordinate (0.0 to 1.0) into pixels
local function calcXcoord(xPercent, xOffset)
    -- (you'd never call this unless you were using Percent)
    local ww = lg.getWidth() -- window width (it might be resizable)
    return (math.floor(ww * xPercent) + xOffset)
end


local function calcYcoord(yPercent, yOffset)
    return (math.floor(lg.getHeight() * yPercent) + yOffset)
end


local function drawDragon()
    ----------------------------------------------------------------------
    -- adjust dragon image Size --
    local ww = lg.getWidth() -- window width (it might be resizable)
    local wh = lg.getHeight()
    local iw = DraImgList.outlines.image:getWidth() -- image width
    local ih = DraImgList.outlines.image:getHeight()

    local wiRatio = (wh / ih) -- window to image ratio (scaling by 'height', the smaller dimension)

    -- how much the images need to be scaled
    local xscale = wiRatio
    local yscale = wiRatio -- just scale dimensions evenly (by width).. no stretching
    ----------------------------------------------------------------------

    -- find offset to Center the image
    -- x-offset needed is the Window center minus the Image center.
    -- How far is the image center from the window center?
    -- draw it that far to the right...
    local xCtr = (ww / 2) - (math.floor(iw / 2 * xscale))
    local yCtr = (wh / 2) - (math.floor(ih / 2 * yscale))

    -----------------------------------------------------------------------
    -- do we want to Shift the image by some Percentage, or by some pixels?
    local xPctShift = 0 -- percent to shift image center (e.g. -0.2 is 20% left)
    local yPctShift = 0

    local xloc = calcXcoord(xPctShift, xCtr + -100) -- final Position of dragon image
    local yloc = calcYcoord(yPctShift, yCtr + 0)
    -----------------------------------------------------------------------


    -- Draw the Dragon parts (in colors)
    lg.setColor(unpack(DraImgList.outlines.color))
    lg.draw(DraImgList.outlines.image, xloc, yloc, 0, xscale, yscale)




    --lg.setColor(unpack(DraImgList.primary.color)) -- old.. Delete
    love.graphics.setColor(CLS.buttonList[iPrimary].color)
    lg.draw(DraImgList.primary.image, xloc, yloc, 0, xscale, yscale)

    --lg.setColor(unpack(DraImgList.secondary.color))
    love.graphics.setColor(CLS.buttonList[iSecondary].color)
    lg.draw(DraImgList.secondary.image, xloc, yloc, 0, xscale, yscale)

    --lg.setColor(unpack(DraImgList.tertiary.color))
    love.graphics.setColor(CLS.buttonList[iTertiary].color)
    lg.draw(DraImgList.tertiary.image, xloc, yloc, 0, xscale, yscale)



    -- draw the Material types (textures)
    --lg.setColor(0, 0, 0) -- (if you wanted "all black" textures)
    local matAlpha = 0.6 -- a value of 0.6 or so lets the underneath show through.
    lg.setColor(1, 1, 1, matAlpha) -- setting to White means the textures will show in their source-file colors.
    lg.draw(DraImgList.primary.matImage, xloc, yloc, 0, xscale, yscale)
    lg.draw(DraImgList.secondary.matImage, xloc, yloc, 0, xscale, yscale)
    lg.draw(DraImgList.tertiary.matImage, xloc, yloc, 0, xscale, yscale)
end


local function drawUI()

    -- update button positions if screen is resized 
    CLSconfig.buttonList[1].x = calcXcoord(.70, 0)
    CLSconfig.buttonList[2].x = calcXcoord(.70, 0)
    CLSconfig.buttonList[3].x = calcXcoord(.70, 0)

    CLSconfig.buttonList[1].y = calcYcoord(.05, 0)
    CLSconfig.buttonList[2].y = calcYcoord(.25, 0)
    CLSconfig.buttonList[3].y = calcYcoord(.45, 0)

    local xPctShift = 0.82 -- percentage coordinate of screen to draw Color list at
    CLSconfig.colorsCanvasData.xPos = calcXcoord(xPctShift, 0)
    CLS.draw()
end


function love.draw() -- Love2D calls this 60 times per second.
    lg.setFont(lg.newFont(18))
    lg.setColor(0, .5, 1)

    if mobile then
        lg.print("Tap to change colors", 20, 35)
        lg.print(verMinutes .. " minutes", 20, 60)
    else
        lg.print("[Space] to randomize colors", 20, 10)
        lg.print("[Esc] to exit", 20, 35)
        -- lg.print(verMinutes .. " minutes", 20, 60)
    end

    drawDragon()
    drawUI()
end


function love.keypressed(key)
    if key == "space" then
        CLS.randomizeButtonColors() -- random colors from the List
        --randomColorDraImgList() -- random colors (Not from the list)
        materialDragonImageList()
    end

    if key == "escape" then
        love.event.quit()
    end

    CLS.keypressed(key)
end


function love.wheelmoved(x, y)
    -- (user code here if desired)

    CLS.wheelmoved(x, y)
end


function love.touchpressed(id, x, y, dx, dy, pressure)
    -- kmk instead, wouldn't it be better to just use mousepressed(istouch) mousereleased(istouch)?
    CLS.randomizeButtonColors() -- random colors from the List
    --randomColorDraImgList() -- random colors (Not from the list)
    materialDragonImageList()
end


function love.mousepressed(x, y, button, istouch, presses)
    -- (user code here if desired)

    CLS.mousepressed(x, y)
end


function love.mousereleased(x, y, button, istouch, presses)
    print("mouse at " .. x, y)

    CLS.mousereleased(x, y)
end


--
--print("\n\n"); error("dke ----- BEAKPOINT -----") -- *********** BREAKPOINT **********
